File: rust/src/charts/heatmap.rs
use crate::normalize::NormalizeResult;

const SHADES: &[char] = &[' ', '░', '▒', '▓', '█'];

fn to_shade(value: f64) -> char {
    let clamped = value.max(0.0).min(1.0);
    let idx = (clamped * (SHADES.len() - 1) as f64).round() as usize;
    SHADES[idx]
}

pub fn render_heatmap(normalized: &NormalizeResult) -> String {
    let data = &normalized.data;
    let num_cols = data.first().map_or(0, |r| r.len());

    if num_cols == 0 {
        return String::new();
    }

    let header = format!(
        "    {}",
        (0..num_cols)
            .map(|i| format!("C{}", i + 1))
            .collect::<Vec<_>>()
            .join(" ")
    );

    let rows: Vec<String> = data
        .iter()
        .enumerate()
        .map(|(row_idx, row)| {
            let cells: String = row
                .iter()
                .map(|&v| to_shade(v).to_string())
                .collect::<Vec<_>>()
                .join(" ");
            format!("R{:02} {}", row_idx + 1, cells)
        })
        .collect();

    let mut result = vec![header];
    result.extend(rows);
    result.join("\n")
}

File: src/utils/heatmap.ts
import { type ChartAnnotations, centerText } from './chart-annotations.js';
import { type NormalizeResult } from './normalize.js';

const SHADES = [' ', '░', '▒', '▓', '█'] as const;

function toShade(value: number): string {
	const clamped = Math.max(0, Math.min(1, value));
	const idx = Math.round(clamped * (SHADES.length - 1));
	return SHADES[idx] ?? ' ';
}

export function renderHeatmap({
	normalized,
	options,
}: {
	normalized: NormalizeResult;
	options?: ChartAnnotations;
}): string {
	const { data } = normalized;
	const numCols = data[0]?.length ?? 0;

	if (numCols === 0) return '';

	const labels = options?.seriesLabels ?? [];
	const header = `    ${Array.from({ length: numCols }, (_, i) => labels[i] ?? `C${i + 1}`).join(' ')}`;
	const rows = data.map((row, rowIdx) => {
		const cells = row.map((v) => toShade(v)).join(' ');
		const rowLabel =
			options?.xLabels?.[rowIdx] ?? `R${String(rowIdx + 1).padStart(2, '0')}`;
		return `${rowLabel.padEnd(3)} ${cells}`;
	});

	const lines = [header, ...rows];

	if (options?.yAxisLabel) {
		lines.unshift(options.yAxisLabel);
	}

	if (options?.xAxisLabel) {
		lines.push(centerText(options.xAxisLabel, header.length));
	}

	return lines.join('\n');
}
